crates/red_knot_python_semantic/src/types.rs

```diff
@@ -471,6 +471,236 @@ impl<'db> Type<'db> {
+                }
+            }
+
+            (Type::None, other) | (other, Type::None) => match other {
+                Type::Never
+                | Type::Any
+                | Type::Unknown
+                | Type::Unbound
+                | Type::Todo
+                | Type::Function(..)
+                | Type::Module(..)
+                | Type::Class(..)
+                | Type::Union(..)
+                | Type::Intersection(..)
+                | Type::Tuple(..) => {
+                    unreachable!("handled above")
+                }
+                Type::IntLiteral(_)
+                | Type::BooleanLiteral(_)
+                | Type::StringLiteral(..)
+                | Type::LiteralString
+                | Type::BytesLiteral(..) => true,
+                Type::None => false,
+                Type::Instance(class_type) => {
+                    !class_type.is_known(db, KnownClass::NoneType) // TODO: is this enough since NoneType is final?
+                }
+            },
+
+            (Type::BooleanLiteral(bool), other) | (other, Type::BooleanLiteral(bool)) => {
+                match other {
+                    Type::Never
+                    | Type::Any
+                    | Type::Unknown
+                    | Type::Unbound
+                    | Type::Todo
+                    | Type::Function(..)
+                    | Type::Module(..)
+                    | Type::Class(..)
+                    | Type::Union(..)
+                    | Type::Intersection(..)
+                    | Type::Tuple(..)
+                    | Type::None => {
+                        unreachable!("handled above")
+                    }
+                    Type::IntLiteral(_)
+                    | Type::StringLiteral(..)
+                    | Type::LiteralString
+                    | Type::BytesLiteral(..) => true,
+                    Type::BooleanLiteral(bool_other) => bool != bool_other,
+                    Type::Instance(class_type) => {
+                        !class_type.is_known(db, KnownClass::Bool) // TODO: is this enough since bool is final?
+                    }
+                }
+            }
+
+            (Type::IntLiteral(int), other) | (other, Type::IntLiteral(int)) => match other {
+                Type::Never
+                | Type::Any
+                | Type::Unknown
+                | Type::Unbound
+                | Type::Function(..)
+                | Type::Module(..)
+                | Type::Class(..)
+                | Type::Todo
+                | Type::Union(..)
+                | Type::Intersection(..)
+                | Type::Tuple(..)
+                | Type::None
+                | Type::BooleanLiteral(..) => {
+                    unreachable!("handled above")
+                }
+                Type::StringLiteral(..) | Type::LiteralString | Type::BytesLiteral(..) => true,
+                Type::IntLiteral(int_other) => int != int_other,
+                Type::Instance(..) => false,
```

Comment:
I'm playing things save here, since
```rs
!class_type.is_known(db, KnownClass::Int)
```
would potentially be wrong here, I assume? There might be a subclass of `int`, in which case instances of that subclass would potentially overlap with an `IntLiteral`.

Reply:
I don't think subclasses of `int` are an issue here. An instance of an `int` subclass cannot be considered an `IntLiteral`; only an object `o` where `o.__class__ == int` can be considered an `IntLiteral`

Reply:
This is spec'd here: https://typing.readthedocs.io/en/latest/spec/literal.html#equivalence-of-two-literals

Reply:
Thanks!